### PR TITLE
Integrate user preferences with local storage

### PR DESCRIPTION
**Pull request by aroralakshya**
#13843 
Signed-Off-By: Lakshay arora <arora.lakshya123@gmail.com>

Before submitting pull request, please check that every commit:

- [ ] Has proper Signed-Off-By
- [ ] Has commit message which describes it
- [ ] Is needed on it's own, if you have just minor fixes to previous commits, you can squash them
- [ ] Any new functionality is covered by tests

### CODE REVIEW

**Comment by nulll-pointer** [comment]
I've given preference to the configuration stored in local storage if the data between local storage and database is inconsistent.

File: js/functions.js [hunk 1]
```diff
@@ -5106,3 +5106,47 @@ jQuery.fn.getPostData = function () {
...
+    } else if(isStorageSupported('localStorage') && $("#login_form").length === 0) {
+        var username = PMA_commonParams.get('user');
+        var server = PMA_commonParams.get('server');
+        var configurations = localStorage.getItem(username + "/" + server);
+        if(typeof configurations === 'undefined' || configurations === null) {
+            configurations = [];
+        } else {
+            configurations = JSON.parse(configurations);
+        }
+        for(i = 0; i < configurations.length; ++i) {
```
Comment (saulotoledo) [inline js/functions.js]: Use `var` to declare the loop variable, or you are creating a global one in your script.

File: js/functions.js [hunk 2]
```diff
@@ -5106,3 +5106,47 @@ jQuery.fn.getPostData = function () {
...
+                    .attr('name', 'configurations')
+                    .attr('value', configurations)
+                    .appendTo('#login_form');
+            }
+        });
+    } else if(isStorageSupported('localStorage') && $("#login_form").length === 0) {
+        var username = PMA_commonParams.get('user');
+        var server = PMA_commonParams.get('server');
+        var configurations = localStorage.getItem(username + "/" + server);
+        if(typeof configurations === 'undefined' || configurations === null) {
```
Comment (saulotoledo) [inline js/functions.js]: Try `if (configurations) {` here.

File: js/functions.js [hunk 3]
```diff
@@ -5106,3 +5106,47 @@ jQuery.fn.getPostData = function () {
...
+            configurations = [];
+        } else {
+            configurations = JSON.parse(configurations);
+        }
+        for(i = 0; i < configurations.length; ++i) {
+            currentConfig = configurations[i];
+            type = currentConfig["type"];
+            if(type === "") break;
```
Comment (saulotoledo) [inline js/functions.js]: I particularly do not like this `break` here. If says that if you have an empty value in the configuration you should stop the full processing. Why that happens? It is not easy to understand just by reading the code. Is there no other way to write this more clearly?

**Review by saulotoledo — CHANGES_REQUESTED** [review]

**Comment by nulll-pointer** [comment]
@saulotoledo, thanks for the suggested changes.

File: js/functions.js [hunk 4]
```diff
@@ -5106,3 +5106,47 @@ jQuery.fn.getPostData = function () {
...
+                    .attr('name', 'configurations')
+                    .attr('value', configurations)
+                    .appendTo('#login_form');
+            }
+        });
+    } else if (isStorageSupported('localStorage') && $("#login_form").length === 0) {
+        username = PMA_commonParams.get('user');
+        server = PMA_commonParams.get('server');
+        var configurations = localStorage.getItem(username + "/" + server);
+        if (typeof configurations === 'undefined' || configurations === null) {
```
Comment (saulotoledo) [inline js/functions.js]: I still prefer just `if (configurations) {` here, but it is ok.

**Review by saulotoledo — CHANGES_REQUESTED** [review]
Yw @nulll-pointer :smile:

I just found some other minor details. Please address them if possible.

In your next commit you can use a `git commit --amend`.

**Comment by saulotoledo** [comment]
@nulll-pointer Ops, I forgot something: there is a codeclimate problem to be solved [here](https://codeclimate.com/github/phpmyadmin/phpmyadmin/pull/14543).

File: js/functions.js (hunk 4 above)
Comment (nulll-pointer) [reply js/functions.js]: Actually, getItem function explicitly returns null if the item is not found.
https://stackoverflow.com/questions/3262605/how-to-check-whether-a-storage-item-is-set

Comment (saulotoledo) [reply js/functions.js]: You are right, but if `configurations` is `null` or `undefined`, the expression `if (configurations)` will return `false`:

```javascript
if (undefined) {
  console.log("returning true");
} else {
  console.log("returning false"); // <- this is the result
}

if (null) {
  console.log("returning true");
} else {
  console.log("returning false"); // <- this is the result
}
```

**Comment by saulotoledo** [comment]
@nulll-pointer There are still some lint errors found by travis, ok? You can find them by clicking on "Details" in the travis link below with a red "x".

**Comment by nulll-pointer** [comment]
Yes, there are lint errors in some other prs as well. Im fixing these one
by one.

**Comment by saulotoledo** [comment]
I will give you my ok, but now you need a review from someone on the project team. @ibennetch, could you point to the right person to do it?

**Review by saulotoledo — APPROVED** [review]
@nulll-pointer I'm giving you my ok, but now you need a review from someone on the project team. @ibennetch, could you point to the right person to do it?

**Pull request closed without merging**